Document Snaps cron jobs feature (#1271)
MetaMask/metamask-docs: snaps/features/cron-jobs.md

---
description: Schedule periodic actions for your users.
sidebar_position: 1
---

# Cron jobs

You can schedule actions to run periodically at fixed times or intervals, also known as "cron jobs."
For example, you can display a dialog or notification in MetaMask at a specific time each day.

## Steps

### 1. Configure a cron job

> To configure a cron job, request the [`endowment:cronjob`](../reference/permissions.md#endowmentcronjob)
> permission specifying a schedule of one or more cron jobs in the `jobs` array.

Reviewer: If a cron job is a scheduled action, is "schedule" redundant here?

```suggestion
To configure a cron job, request the [`endowment:cronjob`](../reference/permissions.md#endowmentcronjob)
permission and define one or more cron jobs in the `jobs` array.
```

> Each job is defined by a [cron expression](https://docs.oracle.com/cd/E12058_01/doc/doc.1014/e12030/cron_expressions.htm)
> and request object that will be sent to the Snap's cron job handler when the job is executed.

Reviewer: This sounds a bit passive, perhaps something like:
```suggestion
Define each job with a [cron expression](https://docs.oracle.com/cd/E12058_01/doc/doc.1014/e12030/cron_expressions.htm) and a request object, which the system sends to the Snap's cron job handler upon execution.
```

For example, to configure a job that executes every minute, add the following to your Snap's manifest file:

```json title="snap.manifest.json"
"initialPermissions": {
  "endowment:cronjob": {
    "jobs": [
      {
        "expression": "* * * * *",
        "request": {
          "method": "execute"
        }
      }
    ]
  }
}
```

### 2. Implement a cron job handler

> Expose an [`onCronjob`](../reference/entry-points.md#oncronjob) entry point, which is triggered at
> the specified schedule with the specified requests defined in the `endowment:cronjob` permission.

Reviewer: ```suggestion
Expose an [`onCronjob`](../reference/entry-points.md#oncronjob) entry point, which is triggered at
the specified schedule with the requests defined in the `endowment:cronjob` permission.
```

The following example handles the `execute` method specified in the previous example:

```typescript title="index.ts"
import type { OnCronjobHandler } from "@metamask/snaps-sdk";

export const onCronjob: OnCronjobHandler = async ({ request }) => {
  switch (request.method) {
    case "execute":
      // Cron jobs can execute any method that is available to the Snap.
      return snap.request({
        method: "snap_dialog",
        params: {
          type: "alert",
          content: panel([
            heading("Cron job"),
            text("This dialog was triggered by a cron job."),
          ]),
        },
      });

    default:
      throw new Error("Method not found.");
  }
};
```

> :::note Access data from cron jobs

Reviewer: nit: A tip perhaps?
